### PR TITLE
Add commit weights and reveal weights functions

### LINKED ISSUE #1939: Commit Reveal Weights
Description:  With the Commit Reveal changes upcoming in subtensor, there is a need to compliment the `set_weights` function with `commit_weights` and `reveal_weights`

see this issue for context on the subtensor side: https://github.com/orgs/opentensor/projects/27/views/9?pane=issue&itemId=58412481

AC:
- implement `commit_weights` function in bittensor
- implement `reveal_weights` function in bittensor
- add relevant tests

### PR DESCRIPTION
**Pull request by opendansor**
### Description of the Change

Add a feature to ```commit``` weights for specific subnet on the Bittensor network and also to ```reveal``` weights for a specific subnet on the Bittensor network. 
 
For Example you can call:
```$ btcli wt commit --netuid 1 --uids 1,2,3,4 --weights 0.1,0.2,0.3,0.4``` 
to commit weights, and call 
``` $ btcli wt reveal --netuid 1 --uids 1,2,3,4 --weights 0.1,0.2,0.3,0.4 ```
to reveal the weights on a network.

### Verification Process

Add test for coverage, and E2E test with a member of the Nucleus team to make sure compatibility on the calls to the blockchain. 

### Release Notes

- Added feature to commit weights for a specific subnet. 
- Added feature to reveal weights for a specific subnet.

### CODE REVIEW

File: tests/integration_tests/test_subtensor_integration.py [hunk 1]
```diff
@@ -22,7 +22,9 @@
 from unittest.mock import MagicMock, patch
 
 import pytest
+import torch
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: Why is this here?

File: tests/integration_tests/test_subtensor_integration.py [hunk 2]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+        )
+        self.subtensor._do_commit_weights = MagicMock(return_value=(True, None))
+
+        success, message = self.subtensor.commit_weights(
+            wallet=self.wallet,
+            netuid=3,
+            uids=uids,
+            weights=weights,
+        )
+        assert success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 3]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+        )
+
+        success, message = self.subtensor.commit_weights(
+            wallet=self.wallet,
+            netuid=1,
+            uids=uids,
+            weights=weights,
+            wait_for_inclusion=True,
+        )
+        assert success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 4]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+
+        self.subtensor.reveal_weights = MagicMock(
+            return_value=(True, "Successfully revealed weights.")
+        )
+        self.subtensor._do_reveal_weights = MagicMock(return_value=(True, None))
+
+        success, message = self.subtensor.reveal_weights(
+            wallet=self.wallet, netuid=3, uids=uids, weights=weights, version_key=0
+        )
+        assert success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 5]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+
+        success, message = self.subtensor.reveal_weights(
+            wallet=self.wallet,
+            netuid=1,
+            uids=uids,
+            weights=weights,
+            version_key=0,
+            wait_for_inclusion=True,
+        )
+        assert success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 6]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+        self.subtensor._do_reveal_weights = MagicMock(return_value=(True, None))
+
+        # Commit weights
+        commit_success, commit_message = self.subtensor.commit_weights(
+            wallet=self.wallet,
+            netuid=3,
+            uids=uids,
+            weights=weights,
+        )
+        assert commit_success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 7]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+
+        # Reveal weights
+        reveal_success, reveal_message = self.subtensor.reveal_weights(
+            wallet=self.wallet,
+            netuid=3,
+            uids=uids,
+            weights=weights,
+            version_key=version_key,
+        )
+        assert reveal_success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 8]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+
+        # Commit weights with wait_for_inclusion
+        commit_success, commit_message = self.subtensor.commit_weights(
+            wallet=self.wallet,
+            netuid=1,
+            uids=uids,
+            weights=weights,
+            wait_for_inclusion=True,
+        )
+        assert commit_success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 9]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+        # Reveal weights with wait_for_inclusion
+        reveal_success, reveal_message = self.subtensor.reveal_weights(
+            wallet=self.wallet,
+            netuid=1,
+            uids=uids,
+            weights=weights,
+            version_key=version_key,
+            wait_for_inclusion=True,
+        )
+        assert reveal_success == True
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 10]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+        )
+
+        success, message = self.subtensor.commit_weights(
+            wallet=self.wallet,
+            netuid=3,
+            uids=uids,
+            weights=weights,
+            wait_for_inclusion=True,
+        )
+        assert success == False
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: tests/integration_tests/test_subtensor_integration.py [hunk 11]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+
+        success, message = self.subtensor.reveal_weights(
+            wallet=self.wallet,
+            netuid=3,
+            uids=uids,
+            weights=weights,
+            version_key=0,
+            wait_for_inclusion=True,
+        )
+        assert success == False
```
Comment (thewhaleking) [inline tests/integration_tests/test_subtensor_integration.py]: use `is` for True/False comparisons

File: bittensor/extrinsics/commit_weights.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+        wait_for_finalization (bool, optional): Waits for the transaction to be finalized on the blockchain.
+        prompt (bool, optional): If ``True``, prompts for user confirmation before proceeding.
+    Returns:
+        Tuple[bool, str]: ``True`` if the weight revelation is successful, False otherwise. And `msg`, a string
+        value describing the success or potential error.
+    This function provides a user-friendly interface for revealing weights on the Bittensor blockchain, ensuring proper
+    error handling and user interaction when required.
+    """
+    if prompt:
+        if not input("Would you like to reveal weights? (y/n) ").lower() == "y":
```
Comment (thewhaleking) [inline bittensor/extrinsics/commit_weights.py]: We use `from rich.prompt.Confirm` for these

File: bittensor/subtensor.py [hunk 13]
```diff
@@ -899,6 +903,269 @@ def make_substrate_call_with_retry():
...
+            who=wallet.hotkey.ss58_address,
+            netuid=netuid,
+            uids=list(uids),
+            values=list(weights),
+            version_key=version_key,
+        )
+
+        _logger.info("Commit Hash: {}".format(commit_hash))
+
+        while retries < max_retries:
```
Comment (thewhaleking) [inline bittensor/subtensor.py:968]: I'm not opposed to this if there's a reason, but is there a reason we're doing this rather than `@retry`?

File: bittensor/subtensor.py [hunk 14]
```diff
@@ -899,6 +903,269 @@ def make_substrate_call_with_retry():
...
+
+        This function allows neurons to reveal their previously committed weight distribution, ensuring transparency
+        and accountability within the Bittensor network.
+        """
+        uid = self.get_uid_for_hotkey_on_subnet(wallet.hotkey.ss58_address, netuid)
+        retries = 0
+        success = False
+        message = "No attempt made. Perhaps it is too soon to reveal weights!"
+
+        while retries < max_retries:
```
Comment (thewhaleking) [inline bittensor/subtensor.py:1089]: I'm not opposed to this if there's a reason, but is there a reason we're doing this rather than @retry?

File: tests/e2e_tests/subcommands/weights/test_commit_weights.py [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+    revealed_weights = subtensor.query_module(
+        module="SubtensorModule", name="Weights", params=[1, uid]  # netuid and uid
+    )
+
+    # Assert that the revealed weights are set correctly
+    assert revealed_weights.value is not None, "Weight reveal not found in storage"
+
+    uid_list = list(map(int, re.split(r"[ ,]+", str(uid))))
+    uids = np.array(uid_list, dtype=np.int64)
+    weight_list = list(map(float, re.split(r"[ ,]+", str(weights))))
```
Comment (thewhaleking) [inline tests/e2e_tests/subcommands/weights/test_commit_weights.py]: As part of Roman's refactoring efforts, we've been trying to get away from `list(map(...`, and instead use comprehensions, so `[int(x) for x in re.split(r"[ ,]+", uid_str)]` for increased readability.

File: bittensor/utils/weight_utils.py [hunk 16]
```diff
@@ -341,3 +346,46 @@ def process_weights_for_netuid(
     bittensor.logging.debug("final_weights", normalized_weights)
 
     return non_zero_weight_uids, normalized_weights
+
+
+def generate_weight_hash(
+    who: str, netuid: int, uids: List[int], values: List[int], version_key: int
```
Comment (thewhaleking) [inline bittensor/utils/weight_utils.py]: Let's not name this `who`. I want to see more consistency in naming. Let's say `ss58_address`, if it's supposed to be the ss58 address.

**Review by thewhaleking — CHANGES_REQUESTED** [review]
Bunch of little things. Overall good.

File: tests/integration_tests/test_subtensor_integration.py [hunk 17]
```diff
@@ -359,6 +361,216 @@ def test_set_weights_failed(self):
...
+        uids = np.array([1, 2, 3, 4], dtype=np.int64)
+        commit_hash = bittensor.utils.weight_utils.generate_weight_hash(
+            who=self.wallet.hotkey.ss58_address,
+            netuid=3,
+            uids=uids.tolist(),
+            values=weights.tolist(),
+            version_key=0,
+        )
+
+        self.subtensor.commit_weights = MagicMock(
```
Comment (basfroman) [inline tests/integration_tests/test_subtensor_integration.py:380]: nitpick: It was decided to use `pytest` without `unittest` if possible. I suggest using `mocker.MagicMock` instead of `uniitest.mock.MagicMock`. I understand that under the hood it's the same thing. But the reference to "mocker" will allow us to get rid of the import of `uniitest.mock`.

The same below.

File: bittensor/commands/weights.py [hunk 18]
```diff
@@ -0,0 +1,257 @@
...
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
```
Comment (basfroman) [inline bittensor/commands/weights.py:18]: pls add explanation as module docstring

File: bittensor/commands/weights.py [hunk 19]
```diff
@@ -0,0 +1,257 @@
...
+            action="store_true",
+            default=False,
+        )
+
+        bittensor.wallet.add_args(parser)
+        bittensor.subtensor.add_args(parser)
+
+    @staticmethod
+    def check_config(config: "bittensor.config"):
+        if not config.is_set("wallet.name") and not config.no_prompt:
```
Comment (basfroman) [inline bittensor/commands/weights.py]: I would suggest swapping the terms. Since Python checks from left to right, it is better to start with lightweight calculations or with parts that have already been calculated.
`if not config.no_prompt and not config.is_set("wallet.name"):`
The same below.

File: bittensor/subtensor.py (hunk 13 above)
Comment (opendansor) [reply bittensor/subtensor.py:968]: There is an `@retry` on the api call itself (LN1011). This `while retries` is happening a level higher, and is present in other functions in subtensor.py which function in a similar manner.

File: bittensor/subtensor.py (hunk 14 above)
Comment (opendansor) [reply bittensor/subtensor.py:1089]: There is an `@retry` on the api call itself (LN1136). This `while retries` is happening a level higher, and is present in other functions in subtensor.py which function in a similar manner.

File: bittensor/commands/weights.py [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+            cli.config.uids = Prompt.ask(f"Enter UIDs (comma-separated)")
+
+        if not cli.config.is_set("weights"):
+            cli.config.weights = Prompt.ask(f"Enter weights (comma-separated)")
+
+        # Parse from string
+        netuid = cli.config.netuid
+        version = bittensor.__version_as_int__
+        uids = np.array(
+            list(map(int, re.split(r"[ ,]+", cli.config.uids))),
```
Comment (basfroman) [inline bittensor/commands/weights.py]: nitpick: list comprehension does not create the object in the middle of processing and bring the code more readable.
`[int(x) for x in re.split(r"[ ,]+", cli_config_uids)]`

the same below

File: bittensor/extrinsics/commit_weights.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of
+# the Software.
+
+# THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
+# THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
```
Comment (basfroman) [inline bittensor/extrinsics/commit_weights.py:18]: pls add module docstring

File: bittensor/extrinsics/commit_weights.py [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+        wait_for_inclusion (bool, optional): Waits for the transaction to be included in a block.
+        wait_for_finalization (bool, optional): Waits for the transaction to be finalized on the blockchain.
+        prompt (bool, optional): If ``True``, prompts for user confirmation before proceeding.
+    Returns:
+        Tuple[bool, str]: ``True`` if the weight commitment is successful, False otherwise. And `msg`, a string
+        value describing the success or potential error.
+    This function provides a user-friendly interface for committing weights to the Bittensor blockchain, ensuring proper
+    error handling and user interaction when required.
+    """
+    if prompt:
```
Comment (basfroman) [inline bittensor/extrinsics/commit_weights.py]: Perhaps it should be combined into one if statement?

File: bittensor/subtensor.py [hunk 23]
```diff
@@ -899,6 +903,269 @@ def make_substrate_call_with_retry():
...
+                wait_for_finalization=wait_for_finalization,
+            )
+
+            if not wait_for_finalization and not wait_for_inclusion:
+                return True, None
+
+            response.process_events()
+            if response.is_success:
+                return True, None
+            else:
```
Comment (basfroman) [inline bittensor/subtensor.py:1040]: nitpick: you don't really need `else` here

The same below

File: tests/e2e_tests/subcommands/weights/test_commit_weights.py (hunk 15 above)
Comment (basfroman) [reply tests/e2e_tests/subcommands/weights/test_commit_weights.py]: +1

File: tests/integration_tests/test_subtensor_integration.py (hunk 17 above)
Comment (RomanCh-OT) [reply tests/integration_tests/test_subtensor_integration.py:375]: +1

File: bittensor/utils/weight_utils.py (hunk 16 above)
Comment (basfroman) [reply bittensor/utils/weight_utils.py]: +1

File: bittensor/subtensor.py (hunk 14 above)
Comment (thewhaleking) [reply bittensor/subtensor.py:1089]: Cool. No worries on this then.

File: tests/integration_tests/test_subtensor_integration.py (hunk 17 above)
Comment (opendansor) [reply tests/integration_tests/test_subtensor_integration.py:380]: In order to get all of `test_subtensor_integration.py` file away from `unittest.mock.MagicMock` will require a refactor of this whole file. This can and should be done, but not in the scope of this PR.

Comment (basfroman) [reply tests/integration_tests/test_subtensor_integration.py:380]: I meant only tests in current PR. Using the mocker fixture does not require additional refactoring or importing.

File: bittensor/subtensor.py (hunk 23 above)
Comment (opendansor) [reply bittensor/subtensor.py:1040]: If you search for `if response.is_success:` in the file, I think this precedent is already set. It also allows us to introduce a specific error later on. 

For example a          
```
 if response.is_success:
      return True
  else:
      raise CommitWeightError(response.error_message)
```

Comment (basfroman) [reply bittensor/subtensor.py:1040]: Perhaps I did not express the thought above quite correctly. I meant that in this case you can write like bellow (without `else`):
```python
            if response.is_success:
                return True, None
            
            return False, response.error_message
```

Comment (opendansor) [reply bittensor/subtensor.py:1040]: Yes, I understood it. 

I have no problem writing the code that way, but it'll be the only function in this file without the `else` statement. I know its a nit-pick here, but this file has a convention and I think for readability it would be worth to sticking to what the standard is throughout the file unless there is a good reason to break it.

Control-F `if response.is_success:` in this file and youll see what I mean.

File: tests/integration_tests/test_subtensor_integration.py (hunk 17 above)
Comment (opendansor) [reply tests/integration_tests/test_subtensor_integration.py:380]: `mocker.MagicMock` does not seem to be a drop in replacement. 

Also, I would stick with the precedent already set in this code. I do think if we move over to another mocking library then it should be a refactor task. 
Think of coming back to this code in 12 months and wondering why test is mocked in one way, but the others are done differently. Why add this complexity if its not warranted?

File: bittensor/commands/weights.py [hunk 24]
```diff
@@ -0,0 +1,283 @@
...
+import os
+import re
+
+import numpy as np
+from rich.prompt import Prompt, Confirm
+import bittensor.utils.weight_utils as weight_utils
+import bittensor
+from . import defaults
+
+"""Module that encapsulates the CommitWeightCommand and the RevealWeightCommand. Used to commit and reveal weights
```
Comment (basfroman) [inline bittensor/commands/weights.py]: pls move this after license and before imports

File: bittensor/extrinsics/commit_weights.py [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+# DEALINGS IN THE SOFTWARE.
+
+from typing import Tuple, List
+from rich.prompt import Confirm
+
+import bittensor
+
+""" Module commit weights and reveal weights extrinsic. """
```
Comment (basfroman) [inline bittensor/extrinsics/commit_weights.py]: pls put this after license, before import

File: tests/integration_tests/test_subtensor_integration.py (hunk 17 above)
Comment (basfroman) [reply tests/integration_tests/test_subtensor_integration.py:380]: There is no complexity in this - pytest is already used in this module.
You've written new tests using an approach that we want to avoid. This was decided by the team.
Thus, technical debt increases.

But since this was just a nitpick, I shared my opinion with you :)

**Review by basfroman — APPROVED** [review]
Small style fix required. Overall looks good!

**Review by thewhaleking — APPROVED** [review]
